hawkbit-repository/hawkbit-repository-jpa/src/main/java/org/eclipse/hawkbit/repository/jpa/JpaDeploymentManagement.java

```diff
@@ -209,10 +235,19 @@ private DistributionSetAssignmentResult assignDistributionSetToTargets(final Lon
         // maximum 1000 elements, so we need to split the entries here and
         // execute multiple statements we take the target only into account if
         // the requested operation is no duplicate of a previous one
-        final List<JpaTarget> targets = Lists.partition(controllerIDs, Constants.MAX_ENTRIES_IN_STATEMENT).stream()
-                .map(ids -> targetRepository
-                        .findAll(TargetSpecifications.hasControllerIdAndAssignedDistributionSetIdNot(ids, set.getId())))
-                .flatMap(List::stream).collect(Collectors.toList());
+        final List<JpaTarget> targets;
+        if (offline) {
```

Comment:
Due to the offline capability there are a lot of branches within the code now. This makes it hard to maintain and you'll get at lot of duplications. Have a look only at the public methods: All of them pass through a boolean to state whether it is an offline or online update (indicator for violating SRP). I suggest to either split this class into two (e.g. `JpaOnlineDeploymentManagement` and `JpaOfflineDeploymentManagement`) or make use of the strategy-pattern for online/offline case e.g.:

```
interface UpdateStrategy {

    List<JpaTarget> findTarget(final List<String> controllerIDs, final long setId);
    void updateTargetStatus(final JpaDistributionSet set, final List<List<Long>> targetIds, final String currentUser);
    Set<Long> findTargetIdToCancel(List<List<Long>> targetIds);
    Status getActionStatus();
}

class OnlineUpdate implements UpdateStrategy{}
class OfflineUpdate implements UpdateStrategy{}
``` 
Later on in the implementation you won't have an if/else statement and only call:
```
strategy.findTarget(ids, set.getId());
```
Instead of:
```
final List<JpaTarget> targets;
        if (offline) {
            targets = Lists.partition(controllerIDs, Constants.MAX_ENTRIES_IN_STATEMENT).stream()
                    .map(ids -> targetRepository.findAll(SpecificationsBuilder.combineWithAnd(Arrays.asList(
                            TargetSpecifications.hasControllerIdAndAssignedDistributionSetIdNot(ids, set.getId()),
                            TargetSpecifications.notEqualToTargetUpdateStatus(TargetUpdateStatus.PENDING)))))
                    .flatMap(List::stream).collect(Collectors.toList());
        } else {
            targets = Lists.partition(controllerIDs, Constants.MAX_ENTRIES_IN_STATEMENT).stream()
                    .map(ids -> targetRepository.findAll(
                            TargetSpecifications.hasControllerIdAndAssignedDistributionSetIdNot(ids, set.getId())))
                    .flatMap(List::stream).collect(Collectors.toList());
        }
```